File: ldb.php
#!/usr/bin/env php
<?php
/**
 * LevelDB Shell
 * Copyright(c) 2012 Shaun Li <shonhen@gmail.com>
 * MIT Licensed
 */

class UsageException extends Exception {
}

class LevelDBShell {

  private $_db;

  public function __construct($dbfolder) {
    $this->_db = new LevelDB($dbfolder);
  }

  protected function db() {
    return $this->_db;
  }

  public function actionKeys($pattern = '') {
    $iter = new LevelDBIterator($this->db());
    foreach ($iter as $key => $value) {
      if (!$pattern)
        echo $key . PHP_EOL;
      if (preg_match('/^' . $pattern . '$/', $key))
        echo $key . PHP_EOL;
    }
  }

  public function actionGet($key) {
    echo $this->db()->get($key) . PHP_EOL;
  }

  public function actionSet($key, $value) {
    $this->db()->set($key, $value);
  }

  public function actionDelete($key) {
    $this->db()->delete($key);
  }

  public static function run($args) {
    if (empty($args))
      throw new UsageException;

    $dbfolder = null;
    $command = null;
    $arguments = array();

    while (!empty($args)) {
      $arg = array_shift($args);
      switch ($arg) {
      case '-d':
        $dbfolder = array_shift($args);
        break;
      default:
        if (!$command)
          $command = $arg;
        else
          $arguments[] = $arg;
        break;
      }
    }

    if (!$dbfolder)
      throw new UsageException('Database folder is missing');

    $shell = new LevelDBShell($dbfolder);
    $method = 'action' . ucfirst($command);
    if (!method_exists($shell, $method))
      throw new UsageException('Unknown command - ' . $command);

    call_user_method_array($method, $shell, $arguments);
  }
}

try {
  LevelDBShell::run(array_slice($argv, 1));
} catch (UsageException $e) {
  if ($msg = $e->getMessage())
    echo $msg . PHP_EOL;
  echo <<< END
Usage:
  ldb -d <folder> <command> [<args>]

The most commonly used commands are:
  keys [pattern]
  set <key> <value>
  get <key>
  delete <key>


END;
}

File: ldb.py
#!/usr/bin/env python
'''
LevelDB Shell
Copyright(c) 2012 Shaun Li <shonhen@gmail.com>
MIT Licensed
'''
from leveldb import LevelDB, WriteBatch
import re
import sys

class UsageException(Exception):
  pass

class LevelDBShell:
  def __init__(self, dbfolder):
    self._db = LevelDB(dbfolder)

  def db(self):
    return self._db

  def actionKeys(self, pattern = None):
    for key, value in self.db().RangeIter():
      if pattern is None:
        print key
      else:
        if re.match(r'^%s$' % pattern, key):
          print key

  def actionGet(self, key):
    print self.db().Get(key)

  def actionSet(self, key, value):
    self.db().Put(key, value)

  def actionDelete(self, key):
    self.db().Delete(key)

  @staticmethod
  def run(args):
    if len(args) == 0:
      raise UsageException()

    dbfolder, command, arguments = None, None, []
    while len(args) > 0:
      arg = args.pop(0)
      if arg == '-d':
        dbfolder = args.pop(0)
      else:
        if command is None:
          command = arg
        else:
          arguments.append(arg)

    if dbfolder is None:
      raise UsageException('Database folder is missing')

    shell = LevelDBShell(dbfolder)
    method = 'action%s%s' % (command[0].upper(), command[1:])
    if not hasattr(shell, method):
      raise UsageException('Unknown command - %s' % command)

    method = getattr(shell, method)
    method(*arguments)

if __name__ == '__main__':
  try:
    LevelDBShell.run(sys.argv[1:])
  except UsageException, e:
    if e.message:
      print e.message
    print '''Usage:
  ldb -d <folder> <command> [<args>]

The most commonly used commands are:
  keys [pattern]
  set <key> <value>
  get <key>
  delete <key>
'''
